exclude_pattern crashes lookout server

## Context

I had some problems when trying to run an analyzer that uses `exclude_pattern` in `FilesRequest` and `ChangesRequest`.

## Steps to reproduce

- clone the lookout-sdk fork

        git clone git@github.com:m09/lookout-sdk.git
- install lookout-sdk (in a venv probably)

        pip install lookout-sdk
- run lookout-sdk server

        python3 language-analyzer.py
- run lookout server

        lookout-sdk review -v ipv4://localhost:2021

## Expected behavior

Analyzer should run with default behavior (comment language and number of functions for each file)

## Observed behavior

Crashes with following log

```
$ lookout-sdk review -v ipv4://localhost:2021
[2018-09-18T10:20:47.88145688+02:00]  INFO resolving to/from references app=lookout-sdk source=lookout-sdk/event.go:55
[2018-09-18T10:20:47.923477623+02:00]  INFO starting a DataServer at ipv4://localhost:10301 app=lookout-sdk source=lookout-sdk/event.go:128
[2018-09-18T10:20:47.923596478+02:00]  INFO starting looking for Analyzer at ipv4://localhost:2021 app=lookout-sdk source=lookout-sdk/event.go:144
[2018-09-18T10:20:47.925695782+02:00]  INFO processing pull request app=lookout-sdk head=HEAD provider= repository=file:///home/mog/work/js/freeCodeCamp source=server/server.go:117
[2018-09-18T10:20:47.92580276+02:00]  INFO getting .lookout.yml app=lookout-sdk head=HEAD provider= repository=file:///home/mog/work/js/freeCodeCamp source=server/server.go:197
[2018-09-18T10:20:47.925897331+02:00]  INFO validating refs: [internal_repository_url:"file:///home/mog/work/js/freeCodeCamp" reference_name:"HEAD" hash:"5cfa1cd78c4db237d050ea70fa41406b343ed430" ], validateRefName: false app=lookout-sdk head=HEAD provider= repository=file:///home/mog/work/js/freeCodeCamp source=git/service.go:36
[2018-09-18T10:20:47.925950972+02:00] DEBUG load trees for references: [{file:///home/mog/work/js/freeCodeCamp HEAD 5cfa1cd78c4db237d050ea70fa41406b343ed430}] app=lookout-sdk head=HEAD provider= repository=file:///home/mog/work/js/freeCodeCamp source=git/service.go:137
[2018-09-18T10:20:47.932146139+02:00]  INFO repository config is not found app=lookout-sdk head=HEAD provider= repository=file:///home/mog/work/js/freeCodeCamp source=server/server.go:216
[2018-09-18T10:20:47.932199406+02:00]  INFO status: pending app=lookout-sdk source=server/server.go:407
[2018-09-18T10:20:47.932252208+02:00]  INFO unary client call started app=lookout-sdk grpc.method=NotifyReviewEvent grpc.service=pb.Analyzer source=grpchelper/logger.go:19 span.kind=client system=grpc
[2018-09-18T10:20:47.934830148+02:00]  INFO streaming server call started app=lookout-sdk grpc.method=GetChanges grpc.service=pb.Data source=grpchelper/logger.go:19 span.kind=server system=grpc
[2018-09-18T10:20:47.934965335+02:00]  INFO validating refs: [internal_repository_url:"file:///home/mog/work/js/freeCodeCamp" reference_name:"HEAD" hash:"ec444f71ad4f4844807d0435de5fc64066367daf"  internal_repository_url:"file:///home/mog/work/js/freeCodeCamp" reference_name:"HEAD" hash:"5cfa1cd78c4db237d050ea70fa41406b343ed430" ], validateRefName: true app=lookout-sdk source=git/service.go:36
[2018-09-18T10:20:47.935014216+02:00] DEBUG load trees for references: [{file:///home/mog/work/js/freeCodeCamp HEAD ec444f71ad4f4844807d0435de5fc64066367daf} {file:///home/mog/work/js/freeCodeCamp HEAD 5cfa1cd78c4db237d050ea70fa41406b343ed430}] app=lookout-sdk source=git/service.go:137
[2018-09-18T10:20:47.960788541+02:00] DEBUG parsing uast for file: common/app/routes/Settings/Settings.jsx app=lookout-sdk source=bblfsh/bblfsh.go:95
[2018-09-18T10:20:47.960889172+02:00]  INFO unary client call started app=lookout-sdk grpc.method=Parse grpc.service=gopkg.in.bblfsh.sdk.v1.protocol.ProtocolService source=grpchelper/logger.go:19 span.kind=client system=grpc
[2018-09-18T10:20:47.962069181+02:00]  INFO streaming client call finished app=lookout-sdk duration=1.165539ms grpc.code=OK grpc.method=Parse grpc.service=gopkg.in.bblfsh.sdk.v1.protocol.ProtocolService grpc.start_time=2018-09-18T10:20:47+02:00 source=grpchelper/logger.go:19 span.kind=client system=grpc
[2018-09-18T10:20:47.962230519+02:00] DEBUG parsing uast for file: common/app/routes/Settings/Settings.jsx app=lookout-sdk source=bblfsh/bblfsh.go:95
[2018-09-18T10:20:47.962295033+02:00]  INFO unary client call started app=lookout-sdk grpc.method=Parse grpc.service=gopkg.in.bblfsh.sdk.v1.protocol.ProtocolService source=grpchelper/logger.go:19 span.kind=client system=grpc
[2018-09-18T10:20:47.962910292+02:00]  INFO streaming client call finished app=lookout-sdk duration=613.537µs grpc.code=OK grpc.method=Parse grpc.service=gopkg.in.bblfsh.sdk.v1.protocol.ProtocolService grpc.start_time=2018-09-18T10:20:47+02:00 source=grpchelper/logger.go:19 span.kind=client system=grpc
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xaf4d86]

goroutine 67 [running]:
github.com/src-d/lookout/service/git.(*regexpFilter).Fn(0xc420f75fb0, 0x0, 0x0, 0x0, 0xc420fe6360)
	/home/mog/go/src/github.com/src-d/lookout/service/git/scanner.go:174 +0x26
github.com/src-d/lookout/service/git.NewChangeFilterScanner.func1(0xc4202d1450, 0xc4202d1450, 0xbc, 0xbc)
	/home/mog/go/src/github.com/src-d/lookout/service/git/scanner.go:217 +0x38
github.com/src-d/lookout.(*FnChangeScanner).Next(0xc420d3d9c0, 0x0)
	/home/mog/go/src/github.com/src-d/lookout/data.go:287 +0x99
github.com/src-d/lookout.(*FnChangeScanner).Next(0xc420d3da00, 0x0)
	/home/mog/go/src/github.com/src-d/lookout/data.go:285 +0x5a
github.com/src-d/lookout.(*FnChangeScanner).Next(0xc420d3da40, 0xc420d05a00)
	/home/mog/go/src/github.com/src-d/lookout/data.go:285 +0x5a
github.com/src-d/lookout.(*FnChangeScanner).Next(0xc420d3da80, 0xc420d05a00)
	/home/mog/go/src/github.com/src-d/lookout/data.go:285 +0x5a
github.com/src-d/lookout/service/bblfsh.(*ChangeScanner).Next(0xc420f459a0, 0x0)
	/home/mog/go/src/github.com/src-d/lookout/service/bblfsh/bblfsh.go:141 +0x45
github.com/src-d/lookout/service/purge.(*purgeChangesScanner).Next(0xc421026240, 0xc421945050)
	/home/mog/go/src/github.com/src-d/lookout/service/purge/purge.go:70 +0x37
github.com/src-d/lookout.(*DataServerHandler).GetChanges(0xc42000cd00, 0xc420d3d780, 0xfda080, 0xc420aa9320, 0x0, 0x0)
	/home/mog/go/src/github.com/src-d/lookout/data.go:85 +0x151
github.com/src-d/lookout/vendor/gopkg.in/src-d/lookout-sdk.v0/pb._Data_GetChanges_Handler(0xc934c0, 0xc42000cd00, 0xfd8d00, 0xc421751c00, 0x0, 0xc420f45130)
	/home/mog/go/src/github.com/src-d/lookout/vendor/gopkg.in/src-d/lookout-sdk.v0/pb/service_data.pb.go:314 +0x10e
github.com/src-d/lookout/util/grpchelper.StreamServerInterceptor.func1(0xc934c0, 0xc42000cd00, 0xfd8e20, 0xc42012d3f0, 0xc421751b40, 0xebea28, 0xc420fc6000, 0x8000)
	/home/mog/go/src/github.com/src-d/lookout/util/grpchelper/logger.go:47 +0x153
github.com/src-d/lookout/vendor/google.golang.org/grpc.(*Server).processStreamingRPC(0xc4201ba380, 0xfdb2e0, 0xc420a44200, 0xc420206200, 0xc420a72390, 0x17cbdc0, 0x0, 0x0, 0x0)
	/home/mog/go/src/github.com/src-d/lookout/vendor/google.golang.org/grpc/server.go:1167 +0x496
github.com/src-d/lookout/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4201ba380, 0xfdb2e0, 0xc420a44200, 0xc420206200, 0x0)
	/home/mog/go/src/github.com/src-d/lookout/vendor/google.golang.org/grpc/server.go:1253 +0x12b1
github.com/src-d/lookout/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4200c79e0, 0xc4201ba380, 0xfdb2e0, 0xc420a44200, 0xc420206200)
	/home/mog/go/src/github.com/src-d/lookout/vendor/google.golang.org/grpc/server.go:680 +0x9f
created by github.com/src-d/lookout/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/home/mog/go/src/github.com/src-d/lookout/vendor/google.golang.org/grpc/server.go:678 +0xa1
```

The log of the analyzer is:

```
$ python language-analyzer.py
starting gRPC Analyzer server at port 2021
got review request internal_id: "a30bce50-d76f-4fe5-ad4d-0f0222bd5a52"
created_at {
  seconds: 1537258847
  nanos: 925684538
}
updated_at {
  seconds: 1537258847
  nanos: 925684604
}
is_mergeable: true
commit_revision {
  base {
    internal_repository_url: "file:///home/mog/work/js/freeCodeCamp"
    reference_name: "HEAD"
    hash: "ec444f71ad4f4844807d0435de5fc64066367daf"
  }
  head {
    internal_repository_url: "file:///home/mog/work/js/freeCodeCamp"
    reference_name: "HEAD"
    hash: "5cfa1cd78c4db237d050ea70fa41406b343ed430"
  }
}
source {
  internal_repository_url: "file:///home/mog/work/js/freeCodeCamp"
  reference_name: "HEAD"
  hash: "5cfa1cd78c4db237d050ea70fa41406b343ed430"
}
merge {
  internal_repository_url: "file:///home/mog/work/js/freeCodeCamp"
  reference_name: "HEAD"
  hash: "5cfa1cd78c4db237d050ea70fa41406b343ed430"
}
configuration {
}

Exception calling application: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Socket closed"
	debug_error_string = "{"created":"@1537258847.968332236","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1099,"grpc_message":"Socket closed","grpc_status":14}"
>
Traceback (most recent call last):
  File "/home/mog/.virtualenvs/lookout-sdk/lib/python3.6/site-packages/grpc/_server.py", line 389, in _call_behavior
    return behavior(argument, context), True
  File "language-analyzer.py", line 43, in NotifyReviewEvent
    for change in changes:
  File "/home/mog/.virtualenvs/lookout-sdk/lib/python3.6/site-packages/grpc/_channel.py", line 363, in __next__
    return self._next()
  File "/home/mog/.virtualenvs/lookout-sdk/lib/python3.6/site-packages/grpc/_channel.py", line 357, in _next
    raise self
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Socket closed"
	debug_error_string = "{"created":"@1537258847.968332236","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1099,"grpc_message":"Socket closed","grpc_status":14}"
>
```

## Environment

- lookout-sdk==0.0.4
- lookout v0.1.3 built on 08-27-2018_12_10_56 __and__ lookout-sdk from master
- Python 3.6.4
- Ubuntu 16.04

--- a/service/git/scanner.go
+++ b/service/git/scanner.go
@@ -171,6 +171,10 @@ func (s *regexpFilter) OnStart() error {
 }
 
 func (s *regexpFilter) Fn(f *lookout.File) (bool, error) {
+	if f == nil {
+		return true, nil
+	}
+
 	if !s.matchInclude(f.Path) {
 		return true, nil
 	}

--- a/service/git/scanner_test.go
+++ b/service/git/scanner_test.go
@@ -148,6 +148,46 @@ func (s *ScannerSuite) TestFilterExcludeOne() {
 	}
 }
 
+// TestFilterDeletedFiles triggers https://github.com/src-d/lookout/issues/261
+func (s *ScannerSuite) TestFilterDeletedFiles() {
+	fixtures := []filterScannerFixture{
+		{},
+		{IncludePattern: ".*"},
+		{ExcludePattern: `\.gitignore`},
+		{IncludePattern: ".*", ExcludePattern: `\.gitignore`},
+	}
+
+	for i, fixture := range s.changeFixtures(fixtures) {
+		head := s.getCommit(s.Basic.Head)
+		headTree, err := head.Tree()
+		s.Require().NoError(err)
+
+		parent, err := head.Parent(0)
+		s.Require().NoError(err)
+		parentTree, err := parent.Tree()
+		s.Require().NoError(err)
+
+		s.T().Run(fmt.Sprintf("change case %d", i), func(t *testing.T) {
+			require := require.New(t)
+
+			cs := NewChangeFilterScanner(
+				// head and parent inverted to force a test case with a deleted file
+				NewDiffTreeScanner(headTree, parentTree),
+				fixture.IncludePattern, fixture.ExcludePattern,
+			)
+
+			var changes []*lookout.Change
+			for cs.Next() {
+				changes = append(changes, cs.Change())
+			}
+
+			require.False(cs.Next())
+			require.NoError(cs.Err())
+			require.NoError(cs.Close())
+		})
+	}
+}
+
 func (s *ScannerSuite) changeFixtures(fs []filterScannerFixture) []*lookout.ChangesRequest {
 	res := make([]*lookout.ChangesRequest, len(fs), len(fs))
 	for i, f := range fs {

